File: Shared/Blog.Core/Attributes/ValidateModelStateAttribute.cs
// <copyright file="ValidateModelStateAttribute.cs" company="PlaceholderCompany">
// Copyright (c) PlaceholderCompany. All rights reserved.
// </copyright>

namespace Blog.Core.Attributes;

using System.Linq;
using Infrastructure.OperationResults;
using Microsoft.AspNetCore.Mvc;
using Microsoft.AspNetCore.Mvc.Filters;

/// <summary>
/// Custom validation handler for availability to whit OperationResult.
/// </summary>
public class ValidateModelStateAttribute : ActionFilterAttribute
{
    /// <inheritdoc cref="ActionNameAttribute"/>
    public override void OnActionExecuting(ActionExecutingContext context)
    {
        if (context.ModelState.IsValid)
        {
            return;
        }

        var operation = OperationResult.CreateResult<object>();
        var messages = context.ModelState.Values.SelectMany(x => x.Errors.Select(xx => xx.ErrorMessage));
        var message = string.Join(" ", messages);
        operation.AddError(message);
        context.Result = new OkObjectResult(operation);
    }
}

File: Shared/Blog.FSharp.Core/Attributes/ValidateModelStateAttribute.fs
namespace Blog.Core.Attributes

open Microsoft.AspNetCore.Mvc
open Microsoft.AspNetCore.Mvc.Filters
open Blog.FSharp.Core.Infrastructure.OperationResults

/// <summary>
/// Custom validation handler for availability to work with OperationResult.
/// This attribute checks model state before the action executes.
/// </summary>
type ValidateModelStateAttribute() =
    inherit ActionFilterAttribute()

    /// <summary>
    /// Called before the action executes. If the model state is invalid,
    /// creates an OperationResult with error messages and returns it.
    /// </summary>
    /// <param name="context">The action executing context.</param>
    override _.OnActionExecuting(context: ActionExecutingContext) =
        if context.ModelState.IsValid then
            () // Do nothing if model state is valid
        else
            // Create an empty OperationResult<obj>
            let operation = OperationResult<obj>()

            // Collect all model state error messages
            let messages =
                context.ModelState.Values
                |> Seq.collect (fun x -> x.Errors |> Seq.map (fun e -> e.ErrorMessage))

            // Concatenate all messages into one string
            let message = String.concat " " messages

            // Add error metadata to the operation result
            OperationResultExtensions.addError operation message |> ignore

            // Set the action result to return the operation result as JSON
            context.Result <- OkObjectResult(operation) :> IActionResult
